aloha-io-avro/src/main/scala/com/eharmony/aloha/audit/impl/avro/AvroGenericRecordAuditor.scala

```diff
@@ -0,0 +1,215 @@
+  require(AuditorSchema != null && ModelIdSchema != null)
+
+  def apply[N](implicit ri: RefInfo[N]): Option[AvroGenericRecordAuditor[N]] = {
+    ri match {
+      case RefInfo.Boolean => opt(IdentityAuditor[Boolean]())
+      case RefInfo.Byte => opt(IdentitySubTypeAuditor[Byte, Int])
+      case RefInfo.Short => opt(IdentitySubTypeAuditor[Short, Int])
+      case RefInfo.Int => opt(IdentityAuditor[Int]())
+      case RefInfo.Long  => opt(IdentityAuditor[Long]())
+      case RefInfo.Float => opt(IdentityAuditor[Float]())
+      case RefInfo.Double => opt(IdentityAuditor[Double]())
+
+      // TODO: Should a String really be inserted to the record or a CharSequence like a avro.util.Utf8?
+      case RefInfo.String => opt(IdentityAuditor[String]())
+      case r if RefInfoOps.isIterable(RefInfo.Boolean, r) => opt(IdentityArrayAuditor[Boolean]())
+      case r if RefInfoOps.isIterable(RefInfo.Byte, r) => opt(IdentitySubTypeArrayAuditor[Byte, Int])
+      case r if RefInfoOps.isIterable(RefInfo.Short, r) => opt(IdentitySubTypeArrayAuditor[Short, Int])
+      case r if RefInfoOps.isIterable(RefInfo.Int, r) => opt(IdentityArrayAuditor[Int]())
+      case r if RefInfoOps.isIterable(RefInfo.Long, r) => opt(IdentityArrayAuditor[Long]())
+      case r if RefInfoOps.isIterable(RefInfo.Float, r) => opt(IdentityArrayAuditor[Float]())
+      case r if RefInfoOps.isIterable(RefInfo.Double, r) => opt(IdentityArrayAuditor[Double]())
+      case r if RefInfoOps.isIterable(RefInfo.String, r) => opt(IdentityArrayAuditor[String]())
+
+      case _ => None
+    }
+  }
+
+  /**
+    * Cast `auditor` to `ScoreAuditor[N]` and wrap in an Option.
+    * '''NOTE''': ''I hate this method'' but I don't want to require an implicit
+    * [[AvroGenericRecordAuditor]] as a parameter in the apply method because [[Impl]]'s
+    * `changeType` method would then need the same implicit to be included.  A determination
+    * should be made as to the feasibility of this strategy.
+    *
+    * @param auditor an auditor to wrap
+    * @tparam N type of auditor to be returned
+    * @return
+    */
+  private[this] def opt[N](auditor: AvroGenericRecordAuditor[_]): Option[AvroGenericRecordAuditor[N]] =
+    Option(auditor.asInstanceOf[AvroGenericRecordAuditor[N]])
+
+  private[this] sealed trait Impl[N] extends AvroGenericRecordAuditor[N] {
+
+    override final def changeType[M: RefInfo]: Option[AvroGenericRecordAuditor[M]] = AvroGenericRecordAuditor[M]
+
+    protected[this] def convertToWire(value: N): Any
+
+    override final def failure(key: ModelIdentity,
+                               errorMsgs: => Seq[String],
+                               missingVarNames: => Set[String],
+                               subValues: Seq[GenericRecord]): GenericRecord = {
+
+      val r = new GenericData.Record(AuditorSchema)
+      r.put("model", mId(key))
+
+      // TODO: Should these be null if iterables are empty?
```

Comment:
I'd go with empty `Iterable`s to show we considered the data and found that they were empty.